platform/linux-generic/include/odp/api/plat/spinlock_recursive_inlines.h

```diff
@@ -0,0 +1,93 @@
+/* Copyright (c) 2013-2018, Linaro Limited
+ * Copyright (c) 2022, Nokia
+ * All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#ifndef _ODP_PLAT_SPINLOCK_RECURSIVE_INLINES_H_
+#define _ODP_PLAT_SPINLOCK_RECURSIVE_INLINES_H_
+
+#include <odp/api/spinlock.h>
+#include <odp/api/thread.h>
+
+#include <odp/api/abi/spinlock_recursive.h>
+
+/** @cond _ODP_HIDE_FROM_DOXYGEN_ */
+
+#ifndef _ODP_NO_INLINE
+	/* Inline functions by default */
+	#define _ODP_INLINE static inline
+	#define odp_spinlock_recursive_init __odp_spinlock_recursive_init
+	#define odp_spinlock_recursive_lock __odp_spinlock_recursive_lock
+	#define odp_spinlock_recursive_trylock __odp_spinlock_recursive_trylock
+	#define odp_spinlock_recursive_unlock __odp_spinlock_recursive_unlock
+	#define odp_spinlock_recursive_is_locked __odp_spinlock_recursive_is_locked
+#else
+	#undef _ODP_INLINE
+	#define _ODP_INLINE
+#endif
+
+_ODP_INLINE void odp_spinlock_recursive_init(odp_spinlock_recursive_t *rlock)
+{
+	odp_spinlock_init(&rlock->lock);
+	rlock->owner = -1;
+	rlock->cnt   = 0;
+}
+
+_ODP_INLINE void odp_spinlock_recursive_lock(odp_spinlock_recursive_t *rlock)
+{
+	int thr = odp_thread_id();
+
+	if (rlock->owner == thr) {
+		rlock->cnt++;
+		return;
+	}
+
+	odp_spinlock_lock(&rlock->lock);
+	rlock->owner = thr;
+	rlock->cnt   = 1;
+}
+
+_ODP_INLINE int odp_spinlock_recursive_trylock(odp_spinlock_recursive_t *rlock)
+{
+	int thr = odp_thread_id();
+
+	if (rlock->owner == thr) {
+		rlock->cnt++;
+		return 1;
+	}
+
+	if (odp_spinlock_trylock(&rlock->lock)) {
+		rlock->owner = thr;
+		rlock->cnt   = 1;
+		return 1;
+	} else {
+		return 0;
+	}
+}
+
+_ODP_INLINE void odp_spinlock_recursive_unlock(odp_spinlock_recursive_t *rlock)
+{
+	rlock->cnt--;
```

Comment:
Need protection for accidental wraparound?

Reply:
Looks like it does (this commit didn't modify the implementation). `odp_spinlock_unlock()` is called only when `rlock->cnt == 0` and calling `odp_spinlock_recursive_unlock()` more times than the lock has been acquired would be against the API spec anyway. The next lock call will reset count to a valid value.

Reply:
Assert for count not being already zero could be helpful in catching any bugs instead of program just getting stuck.

Reply:
The problem with asserts is that we cannot currently use them in inline headers as they are not part of the public API. We should probably do something about this limitation in a separate PR.

Reply:
OK, resolving this.